### PR TITLE
[BIOMAGE-1459] Add workaround to cypress bug

### PR DESCRIPTION
**Pull request by cosa65**
# Background
#### Link to issue 
https://biomage.atlassian.net/browse/BIOMAGE-1459

#### Anything else the reviewers should know about the changes here
There is a bug with the cypress testing library and it is affecting us. You can check [this thread](https://github.com/cypress-io/cypress/issues/2831) on the problem (last message 9 days ago).

Due to the bug when a test fails the `afterEach` code errors out, this causes us to have some problems when we want to, for example, create a project in `beforeEach` and delete it in `afterEach`.

The workaround that I decided to go with is running the `afterEach` as a small test that runs `cleanup` (which contains all the code that should have gone into `afterEach`. 

The downside to this workaround is that these cleanup tests also run the `beforeEach` so we can't really use `beforeEach` for everything we would like to. However, given the situation, I think replacing a bugged `afterEach` that will cause problems across runs of integration testing is probably the best option, even if that limits our use of `beforeEach` too.

UPDATED after discussing with pol to inserting the cleanup at tests end (and in the `beforeEach` too).
 
# Changes
### Code changes

# Definition of DONE
Your changes will be ready for merging after each of the steps below have been completed:

### Testing
- [ ] Unit tests written
- [x] Tested locally with Inframock (with latest production data downloaded with biomage experiment pull)
- [ ] Deployed to staging

To set up easy local testing with inframock, follow the instructions here: https://github.com/biomage-ltd/inframock
To deploy to the staging environment, follow the instructions here: https://github.com/biomage-ltd/biomage-utils

### Documentation updates
Is all relevant documentation updated to reflect the proposed changes in this PR?

- [ ] Relevant Github READMEs updated
- [ ] Relevant wiki pages created/updated

### Approvers
- [ ] Approved by a member of the core engineering team
- [ ] (UX changes) Approved by vickymorrison (this is her username, tag her if you need approval)

### Just before merging:
- [ ] After the PR is approved, the `unstage` script in here: https://github.com/biomage-ltd/biomage-utils is executed. This script cleans up your deployment to staging

### Optional
- [ ] Photo of a cute animal attached to this PR

### CODE REVIEW

**Review by kafkasl — COMMENTED** [review]
I don't think this is a good solution, it goes against the best practices recommended by cypress. They basically recommend to do cleanups before actually running a test not after it. You can get more info here: https://docs.cypress.io/guides/references/best-practices#Using-after-or-afterEach-hooks. It's pretty well explained and it makes a lot of sense so I think I'd go with that approach.

**Comment by cosa65** [comment]
> I don't think this is a good solution, it goes against the best practices recommended by cypress. They basically recommend to do cleanups before actually running a test not after it. You can get more info here: https://docs.cypress.io/guides/references/best-practices#Using-after-or-afterEach-hooks. It's pretty well explained and it makes a lot of sense so I think I'd go with that approach.

In our case that would mean deleting the project and creating it again, right? 
So we woudl have something like this in the `beforeEach`:
`cy.deleteProject();`
`cy.createProject();`
(*)

That was another solution we thought about, the problem with that one is that after the last test we would be leaving a test project created lying around in staging/production. 

In the case that we cleaned up this project after all integration testing finishes we would then need to have `deleteProject` handle the case in which the project exists and the case in which it doesn't (so that it doesn't fail in the first test that runs (*)).

So `deleteProject` would need to have something like [this](https://docs.cypress.io/api/events/catalog-of-events#To-conditionally-turn-off-uncaught-exception-handling-unhandled-promise-rejections) running if a flag is set which is pretty ugly too, but I believe having a test project lying around (and worse, having the integration testing fail if someone deletes this project for some reason) is probably even worse.

**Comment by kafkasl** [comment]
I think that cleaning up before the create project sounds good. Ideally, we could have a `deleteProjectIfExist` functionality testing if it's needed to delete it (otherwise the exception works too). Moreover, after reading some best practices there I think we might want to merge some of the tests into a single one (not in this PR) so we probably would need to do that "delete project" once (because only one test would create the project).

File: e2e/cypress/support/commands.js [hunk 1]
```diff
@@ -71,7 +71,7 @@ Cypress.Commands.add('createProject', (projectName, projectDescription) => {
   });
 
   log.snapshot('open-modal');
-  cy.get('[data-test-id="create-new-project-button"]').click({ force: true });
+  cy.get('[data-test-id="create-new-project-button"]').scrollIntoView().click();
```
Comment (kafkasl) [inline e2e/cypress/support/commands.js:74]: why this change? I don't remember why I had to use the `force: true` but the modal shouldn't need any scrolling right?

File: e2e/cypress/support/commands.js [hunk 2]
```diff
@@ -265,3 +265,19 @@ Cypress.Commands.add('waitForQc', (timeout, numQcSteps = 7) => {
...
   log.end();
 });
+
+Cypress.Commands.add('cleanUpProjectIfNecessary', (projectName) => {
+  const log = Cypress.log({
+    displayName: 'Cleaning up previous project',
+    message: ['clean project'],
+  });
+
+  const projectExists = Cypress.$(`span:contains(${projectName})`).length;
```
Comment (kafkasl) [inline e2e/cypress/support/commands.js:275]: I think the selector here should include `data-test-project-card` to make it more clear and robust. Maybe you can just do `data-test-project-card.:contains(${projectName})`, otherwise you might need to do  `data-test-project-card > span.:contains(${projectName})`. Essentially, I'd prefer to remove the `span` if possible, because it couples the HTML tag to the test. The test then would break if we ever change the HTML element as opposed to just using the `data-test-project-card`  which should never change.

File: e2e/cypress/integration/ui-smoke/ui-add-sample.spec.js [hunk 3]
```diff
@@ -42,13 +47,15 @@ describe('Sample addition/removal', () => {
...
+
+    cy.cleanUpProjectIfNecessary(projectName);
   });
 
   it('Adds a new sample correctly by input selection', () => {
     cy.addSample(addFileActions.SELECT_INPUT);
 
     // Sample cell shows up
-    cy.contains('.data-test-sample-in-table-name', 'WT1', { timeout: 10000 }).should('be.visible');
+    cy.contains('.data-test-sample-in-table-name', 'WT1', { timeout: 10000 }).should('exist');
```
Comment (kafkasl) [inline e2e/cypress/integration/ui-smoke/ui-add-sample.spec.js:58]: why this change?

**Review by kafkasl — APPROVED** [review]
